src/Hedgehog/Internal/Function.idr

```diff
@@ -0,0 +1,351 @@
+
+export
+Cogen a => Cogen (Maybe a) where
+  perturb Nothing  = variant 0
+  perturb (Just x) = perturb x . shiftArg . variant 1
+
+export
+Cogen a => Cogen (List a) where
+  perturb []      = variant 0
+  perturb (x::xs) = perturb xs . shiftArg . perturb x . shiftArg . variant 1
+
+Integral a => Neg a => Ord a => Cast a (Bool, List Bool) where
+  cast n = if n >= 0 then (True, go [] n) else (False, go [] $ -n - 1) where
+    go : List Bool -> a -> List Bool
+    go bits x = if x == 0 then bits else go ((mod x 2 == 1) :: bits) (assert_smaller x $ div x 2)
+
+Integral a => Neg a => Cast (Bool, List Bool) a where
+  cast (sign, bits) = do
+    let body = foldl (\acc, b => acc * the a 2 + if b then 1 else 0) (the a 0) bits
+    if sign then body else negate $ body + 1
+
+export
+Integral a => Neg a => Ord a => Cogen a where
+  perturb = perturb . cast {to=(Bool, List Bool)}
+
+export
+Cogen String where
+  perturb = perturb . fastUnpack
+
+namespace Derive
+
+  ||| Derivation facility for `Gogen` interface
+  |||
+  ||| Use `derive`, `deriveIndexed` or `derivePattern` from `Language.Reflection.Derive`
+  ||| for simple, purely indexed or mixed data types.
+  export
+  CogenVis : Visibility -> List Name -> ParamTypeInfo -> Res (List TopLevel)
+  CogenVis vis nms p = do
+    let fun  := funName p "perturb"
+    let impl := implName p "Cogen"
+    Right [ TL (perturbClaim fun p) (perturbDef fun p.info)
+          , TL (cogenImplClaim impl p) (cogenImplDef fun impl)
+          ]
+    where
+      perturbClaim : Name -> ParamTypeInfo -> Decl
+      perturbClaim fun p = simpleClaim vis fun $ piAll `(~(p.applied) -> Seed -> Seed) $ allImplicits p "Cogen"
+
+      perturbDef : Name -> TypeInfo -> Decl
+      perturbDef fun ti = def fun $ map clause $ [0 .. length ti.cons] `zip` ti.cons where
+        clause : (Nat, Con _ _) -> Clause
+        clause (idx, con) = accumArgs unerased (\x => `(~(var fun) ~x)) rhs arg con where
+
+          arg : BoundArg 1 Unerased -> TTImp
+          arg $ BA g [x] _ = assertIfRec nms g.type `(Function.perturb ~(varStr x) . Function.shiftArg)
+
+          rhs : SnocList TTImp -> TTImp
+          rhs = foldr (\l, r => `(~l . ~r)) `(Seed.variant ~(primVal $ B64 $ cast idx))
+
+      cogenImplClaim : Name -> ParamTypeInfo -> Decl
+      cogenImplClaim impl p = implClaimVis vis impl $ implType "Cogen" p
+
+      cogenImplDef : (fun, impl : Name) -> Decl
+      cogenImplDef fun impl = def impl $ pure $ patClause (var impl) (var "MkCogen" `app` var fun)
+
+
+  ||| Alias for `CogenVis Public`
+  export %inline
+  Cogen : List Name -> ParamTypeInfo -> Res (List TopLevel)
+  Cogen = CogenVis Public
+
+||| Generates a random function being given a generator of codomain type
+|||
+||| This function takes a co-generator of domain type using `auto`-argument based on the type.
+||| This generator does not shrink.
+|||
+||| Notice that this generator returns a non-showable value (unless you invent your own implementation).
```

Comment:
Please keep all lines (docs and code) at 80 or fewer characters unless strictly necessary otherwise.

Reply:
Oh, really? I'll do so, if you insist, but do you really use such a narrow monitor? I thought that already a dozen of years standard length shifted to at least 120 characters.

Reply:
I insist. It's the limit I use in all my projects. This is not about narrow monitors but about readability and using split screens a lot. Here are a couple of projects / organisations who adhere to 80 characters width:

[Jonathan Tibbel](https://github.com/tibbe/haskell-style-guide/blob/master/haskell-style.md)
[A CS Haskell course](https://www.cs.umd.edu/class/spring2022/cmsc488B/style.html)
[Another CS Haskell course; 78 characters only](http://courses.cms.caltech.edu/cs11/material/haskell/misc/haskell_style_guide.html)
[The official Scala style guide](https://docs.scala-lang.org/style/indentation.html)

Reply:
Done